Volodymyr sakharevych

vsakharevych opened the pull request:
Please have a look at solutions for tasks 1 & 2 (implementations of dynamic list of c-strings & class AnyType)

vsakharevych committed:
completed task

vsakharevych committed:
completed task 2

vsakharevych committed:
corrected shortcomings indicated by reviewers

vsakharevych committed:
small adjustments

vsakharevych committed:
restructured AnyType class

vsakharevych committed:
some amendments

vsakharevych committed:
* added solution for task 3 (code analyzer which uses threads pool to process files)
* carried out small refactoring of class AnyType

MarsLviv commented on StringList/helperFunctions.cpp:
```diff
@@ -0,0 +1,88 @@
+
+char* InputString()
+{
+    char* buf = static_cast<char*>(malloc(sizeof(char) * 100));
```
malloc returning a pointer or null. you need check it

vsakharevych replied:
Fixed that

MarsLviv commented on StringList/StringList.cpp:
```diff
@@ -0,0 +1,357 @@
+    {
+        // adding first node
+        list[1] = static_cast<char*>
+                (malloc(sizeof(char**) * 2));
```
malloc returning a pointer or null. you need check it

vsakharevych replied:
It is checked in line 63:
if(!list[1]) 
   ///

MarsLviv commented on StringList/StringList.cpp:
```diff
@@ -0,0 +1,357 @@
+void StringListInit(char*** list)
+{
+    // creating dummy node
+    *list = static_cast<char**>(malloc(sizeof(char**) * 2));
```
malloc returning a pointer or null. you need check it

vsakharevych replied:
It is checked in line 12:
if (*list == nullptr)
        exit(-1);

vsakharevych committed:
added check if files deque is empty

vsakharevych committed:
removed odd try-catch block, added comments

vsakharevych committed:
eliminated bugs indicated by reviewer; added comments; changed strategy of handling exceptional situations

vsakharevych committed:
allocated responsibilities to browse dirs & analyze files between two separate classes; refactored code

vsakharevych committed:
corrected mistake in menu

FedorenkoAndriy commented on StringList/StringList.cpp:
```diff
@@ -0,0 +1,383 @@
+    cur[1] = nullptr;
+}
+
+void StringListDestroy(char*** list)
```
This function is too complicated

FedorenkoAndriy commented on StringList/StringList.cpp:
```diff
@@ -0,0 +1,383 @@
+        list[1] = static_cast<char*>
+                (malloc(sizeof(char**) * 2));
+
+        if (!list[1])
```
Please avoid code duplicates(see line 82-87) move it to separate function

FedorenkoAndriy commented on StringList/StringList.cpp:
```diff
@@ -0,0 +1,383 @@
+
+void StringListRemove(char** list, String str)
+{
+    if (!list[1])
```
The same code duplicate

FedorenkoAndriy commented on StringList/helperFunctions.cpp:
```diff
@@ -0,0 +1,109 @@
+    if(!input)
+    {
+        cout << "Failed to allocate memory.\n";
+        free(buf);
```
If you are checking to null before free you should do it everywhere or nowhere

vsakharevych committed:
Brought project into compliance with expert's remarks: added separate functions for
* memory allocation & check
* deleting nodes
* checking list for emptiness
* arguments check
* replacement of substring in string
Made sure free isn't called on null

vsakharevych committed:
•	deleted excessive code in function StringListDestroy
•	removed excessive header in file stringlist.cpp
•	simplified function StringListRemove
